decompiler/pipeline/controlflowanalysis/loop_utility_methods.py

```diff
@@ -210,3 +228,89 @@ def _requirement_without_reinitialization(ast: AbstractSyntaxTree, node: Abstrac
+        ):
+            return None
+
+        _unify_binary_operation_in_assignment(continuation.instruction)
+        equalizable_nodes.append(code_node)
+    return equalizable_nodes
+
+
+def _is_expression_simple_binary_operation(expression: Expression) -> bool:
+    """Checks if an expression is a simple binary operation. Meaning it includes a variable and a constant and uses plus or minus as operation type."""
+    return (
+        isinstance(expression, BinaryOperation)
+        and expression.operation in {OperationType.plus, OperationType.minus}
+        and any(isinstance(operand, Constant) for operand in expression.subexpressions())
+        and any(isinstance(operand, Variable) for operand in expression.subexpressions())
+    )
+
+
+def _get_variable_in_binary_operation(binaryoperation: BinaryOperation) -> Variable:
+    """Returns the used variable of a binary operation if available."""
+    for operand in binaryoperation.subexpressions():
+        if isinstance(operand, Variable):
+            return operand
+    return None
+
+
+def _count_unaryoperations_negations(expression: Expression) -> int:
+    """Counts the amount of UnaryOperation negations of an expression."""
+    negations = 0
+    for subexpression in expression.subexpressions():
+        if isinstance(subexpression, UnaryOperation) and subexpression.operation == OperationType.negate:
+            negations += 1
+    return negations
+
+
+def _unify_binary_operation_in_assignment(assignment: Assignment):
+    """Brings a simple binary operation of an assignment into a unified representation like 'var = -var + const' instead of 'var = const - var'."""
+    if not assignment.value.operation == OperationType.plus:
+        assignment.substitute(assignment.value, BinaryOperation(OperationType.plus, [assignment.value.left, assignment.value.right]))
+        assignment.substitute(assignment.value.right, UnaryOperation(OperationType.negate, [assignment.value.right]))
+
+    if any(isinstance(operand, Constant) for operand in assignment.value.left.subexpressions()):
+        assignment.substitute(assignment.value, BinaryOperation(OperationType.plus, [assignment.value.right, assignment.value.left]))
+
+
+def _substract_continuation_from_last_definition(code_node: CodeNode, continuation: AstInstruction, variable_init: AstInstruction):
+    """
+    Substracts the value of the continuation instruction from the last definition, which must be a simple binary operation or a constant,
+    defining the same value as the continuation instruction in the given code node.
+
+    :param code_node: Code node whose last instruction is to be changed
+    :param continuation: Instruction defining the for-loops modification
+    :param variable_init: Instruction defining the for-loops declaration
+    """
+    last_definition = code_node.instructions[_get_last_definition_index_of(code_node, variable_init.instruction.destination)]
```

Comment:
Again, why the initialization. And all this recompilation is inefficient. Besides, I am not sure whether this is enough. What if it is not the last instruction, and the continuation is `a = b + 5`, but `b` is defined after this instruction in the block. Does this function check this too?

Reply:
First part fixed in 29ec9fa

Reply:
This isn't relevant anymore since 795c68e. Defined and used variables has to be the same.